# Assignment 2: Entity Resolution (Part 2)


## Objective

In Assignment 2 (Part 2), you will learn how to use Active Learning to address the entity resolution problem. After completing this assignment, you should be able to answer the following questions:

1. Why Active Learning?
2. How to implement uncertain sampling, a popular query strategy for Active Learning?
3. How to solve an ER problem using Active Learning?


## Active Learning

[Active learning](http://tiny.cc/al-wiki) is a certain type of ML algorithms that can train a high-quality ML model with small data-labeling cost. Its basic idea is quite easy to understand. Consider a typical supervised ML problem, which requires a (relatively) large training dataset. In the training dataset, there may be only a small number of data points that are beneficial to the trained ML model. In other words, labeling a small number of data points is enough to train a high-quality ML model. The goal of active learning is to help us to identify those data points. 


In this assignment, we will develop an Active Learning approach for Entity Resolution. The following figure shows the architecture of an entity resolution solution. It consists of four major steps. **I will provide you the source code for Steps 1, 2, 4. Your job is to implement Step 3.**  

<img src="img/arch.png", width=800/>

### Step 1. Read Data

Suppose we get a restaurant dataset `restaurant.csv`. The data has many duplicate restaurants.  For example, the first two rows shown below are duplicated (i.e., refer to the same real-world entity). You can check out all duplicate (matching) record pairs from `true_matches.json`. 

In [1]:
import pandas as pd

df = pd.read_csv('restaurant.csv')
data = df.values.tolist()
print("(#Rows, #Cols) :", df.shape)
df.head(5)

(#Rows, #Cols) : (858, 5)


id                       name                  address         city  \
0   1  arnie morton's of chicago   435 s. la cienega blv.  los angeles   
1   2  arnie morton's of chicago  435 s. la cienega blvd.  los angeles   
2   3         art's delicatessen      12224 ventura blvd.  studio city   
3   4                 art's deli      12224 ventura blvd.  studio city   
4   5              hotel bel-air     701 stone canyon rd.      bel air   

          type  
0     american  
1  steakhouses  
2     american  
3        delis  
4  californian

### Step 2. Similar Pairs

We first use a similarity-join algorithm to generate similar pairs. 

Below is the code. After running the code, we get 678 similar pairs ordered by their similarity decreasingly.

In [2]:
from a2_utils import *

data = df.values.tolist()
simpairs = simjoin(data)

print("Num of Pairs: ", len(data)*(len(data)-1)/2)
print("Num of Similar Pairs: ", len(simpairs))
print("The Most Similar Pair: ", simpairs[0])

Num of Pairs:  367653.0
Num of Similar Pairs:  678
The Most Similar Pair:  ([170, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern/soul'], [169, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern'])


We can see that `simjoin` helps us remove the number of pairs from 367653 to 678. But, there are still many non-matching pairs in `simpairs` (see below). 

In [3]:
print(simpairs[-1])

([764, "buzio's in the rio", '3700 w. flamingo rd.', 'las vegas', 'seafood'], [542, 'carnival world', '3700 w. flamingo rd.', 'las vegas', 'buffets'])


Next, we will use active learning to train a classifier, and then use the classifier to classify each pair in `simpairs` as either "matching" or "nonmatching". 

### Step 3. Active Learning

Given a set of similar pairs, what you need to do next is to iteratively train a classifier to decide which pairs are truly matching. We are going to use [logistic regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) as our classifier. 

#### Initialization

At the beginning, all the pairs are unlabeled. To initialize a model, we first pick up ten pairs and then label each pair using  the `crowdsourcing()` function. You can assume that `crowdsourcing()` will ask a crowd worker (e.g., on Amazon Mechanical Turk) to label a pair. 


`crowdsourcing(pair)` is a function that simulates the use of crowdsourcing to label a pair
  
  - **Input:**	pair – A pair of records 

  - **Output:**	Boolean –  *True*: The pair of records are matching; *False*: The pair of records are NOT matching;

Please use the following code to do the initialization. 

In [4]:
from a2_utils import crowdsourcing

# choose the most/least similar five pairs as initial training data
init_pairs = simpairs[:5] + simpairs[-5:]
# print(simpairs[:2])
# print(init_pairs)
matches = []
nonmatches = []
for pair in init_pairs:
    is_match = crowdsourcing(pair)
    if is_match == True:
        matches.append(pair)
    else:
        nonmatches.append(pair)
        
print("Number of matches: ", len(matches))
print("Number of nonmatches: ", len(nonmatches))

Are they matching?
[170, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern/soul']
[169, "mary mac's tea room", '224 ponce de leon ave.', 'atlanta', 'southern']
Answer: Yes
Are they matching?
[88, 'manhattan ocean club', '57 w. 58th st.', 'new york city', 'seafood']
[87, 'manhattan ocean club', '57 w. 58th st.', 'new york', 'seafood']
Answer: Yes
Are they matching?
[112, 'san domenico', '240 central park s.', 'new york city', 'italian']
[111, 'san domenico', '240 central park s', 'new york', 'italian']
Answer: Yes
Are they matching?
[197, 'fleur de lys', '777 sutter st.', 'san francisco', 'french (new)']
[196, 'fleur de lys', '777 sutter st.', 'san francisco', 'french']
Answer: Yes
Are they matching?
[8, 'cafe bizou', '14016 ventura blvd.', 'sherman oaks', 'french bistro']
[7, 'cafe bizou', '14016 ventura blvd.', 'sherman oaks', 'french']
Answer: Yes
Are they matching?
[709, 'arcadia', '21 e. 62nd st.', 'new york city', 'american (new)']
[66, 'four seasons', '99 e. 5

Here is the only code you need to write in this assignment.


In [17]:
from a2_utils import featurize, crowdsourcing
from sklearn.linear_model import LogisticRegression
from functools import reduce 
import numpy as np

labeled_pairs = matches + nonmatches
unlabeled_pairs = [p for p in simpairs if p not in labeled_pairs]

# Let Match = 1, Not Match = 0
target_labels = [1,1,1,1,1,0,0,0,0,0]

iter_num = 5
for i in range(iter_num):
    print("len(labeled_pairs): ",len(labeled_pairs))
    # Apply Featurize function to each pair (in both labeled_pairs and unlabeled_pairs)
    featurized_labeled_pairs = list(map(featurize, labeled_pairs))
    featurized_unlabeled_pairs = list(map(featurize, unlabeled_pairs))

    model = LogisticRegression()
    # Train labeled_pairs
    model.fit(featurized_labeled_pairs,target_labels)
    
    # Get prediction probability from predicting the unlabled pairs (probability where there are pair column of Match and Non-Match)
    prediction_proba = model.predict_proba(featurized_unlabeled_pairs)
    
    # Flatten the prediction_proba
    flatten_list_proba = reduce(lambda l1,l2: l1+l2, prediction_proba.tolist() )

    # The most uncertain value means that we are not quite sure about Match(1) or Not Match(0)
    # so the probability will be around 0.5 
    uncertain_val = 0.5
    
    # Select only some relevant element (get only Even index as we just want some value close to 0.5 whether it is Match or Not Match)
    flatten_list_proba = [v for i, v in enumerate(flatten_list_proba) if i % 2 == 0]
    
    # Select value that is close to 0.5 (uncertain value)
    close_to_uncertain_val  = min(flatten_list_proba, key=lambda x:abs(x-uncertain_val))
    print("close_to_uncertain val:", close_to_uncertain_valuncertain_val)
    # Get index of the most uncertain value 
    index_closest_to_uncertain_val = flatten_list_proba.index(close_to_uncertain_val)
    
    uncertain_unlabeled_pairs = unlabeled_pairs[index_closest_to_uncertain_val]
    
    # Call crowdsourcing function to help label the uncertain unlabeled pairs
    is_match = crowdsourcing(uncertain_unlabeled_pairs)
    if is_match == True:
        target_labels.append(1)
    else:
        target_labels.append(0)
        
    # Add that pair to labeled_pair list
    labeled_pairs.append(uncertain_unlabeled_pairs)
    # Remove that pair from unlabeled_pair list
    unlabeled_pairs.pop(index_closest_to_uncertain_val)
    print("Target labels: ",target_labels)
    print("Number of labeled_pairs: ",len(labeled_pairs))
    print("Number of unlabeled_pairs: ",len(unlabeled_pairs))

len(labeled_pairs):  10


D:\Programs\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


close_to_uncertain val: 0.49664193820061564
Are they matching?
[741, 'palm too', '840 second ave.', 'new york city', 'steakhouses']
[740, 'palm', '837 second ave.', 'new york city', 'steakhouses']
Answer: No
Target labels:  [1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0]
Number of labeled_pairs:  11
Number of unlabeled_pairs:  667
len(labeled_pairs):  11


D:\Programs\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


close_to_uncertain val: 0.49500970536959943
Are they matching?
[217, 'ritz-carlton dining room (san francisco)', '600 stockton st.', 'san francisco', 'french (new)']
[216, 'ritz-carlton restaurant and dining room', '600 stockton st.', 'san francisco', 'american']
Answer: Yes
Target labels:  [1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1]
Number of labeled_pairs:  12
Number of unlabeled_pairs:  666
len(labeled_pairs):  12


D:\Programs\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


close_to_uncertain val: 0.4974012126266141
Are they matching?
[393, 'le colonial', '149 e. 57th st.', 'new york', 'asian']
[392, 'le chantilly', '106 e. 57th st.', 'new york', 'french']
Answer: No
Target labels:  [1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0]
Number of labeled_pairs:  13
Number of unlabeled_pairs:  665
len(labeled_pairs):  13


D:\Programs\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


close_to_uncertain val: 0.5007639843318441
Are they matching?
[84, 'lespinasse (new york city)', '2 e. 55th st.', 'new york city', 'asian']
[83, 'lespinasse', '2 e. 55th st.', 'new york', 'american']
Answer: Yes
Target labels:  [1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1]
Number of labeled_pairs:  14
Number of unlabeled_pairs:  664
len(labeled_pairs):  14


D:\Programs\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


close_to_uncertain val: 0.5050552086444402
Are they matching?
[150, "bone's restaurant", '3130 piedmont rd. ne', 'atlanta', 'steakhouses']
[149, "bone's", '3130 piedmont road', 'atlanta', 'american']
Answer: Yes
Target labels:  [1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1]
Number of labeled_pairs:  15
Number of unlabeled_pairs:  663


**[Algorithm Description].**   Active learning has many [query strategies](http://tiny.cc/al-wiki-qs) to decide which data point should be labeled. You need to implement uncertain sampling. The algorithm trains an initial model on `labeled_pairs`. Then, it iteratively trains a model. At each iteration, it first applies the model to `unlabeled_pairs`, and makes a prediction on each unlabeled pair along with a probability, where the probability indicates the confidence of the prediction. After that, it selects the most uncertain pair (If there is still a tie, break it randomly),  and call the `crowdsroucing()` function to label the pair. After the pair is labeled, it updates `labeled_pairs` and `unlabeled_pairs`, and then retrain the model on `labeled_pairs`.

**[Input].** 
- `labeled_pairs`: 10 labeled pairs (by default)
- `unlabeled_pairs`: 668 unlabeled pairs (by default)
- `iter_num`: 5 (by default)

**[Output].** 
- `model`: A logistic regression model built by scikit-learn


### Step 4. Model Evaluation

After training an model, you can use the following code to evalute it.

In [16]:
import json
from a2_utils import evaluate

            
sp_features = np.array([featurize(sp) for sp in simpairs])
label = model.predict(sp_features)
pair_label = zip(simpairs, label)

identified_matches = []
for pair, label in pair_label:
    if label == 1:
        identified_matches.append(pair)
        
precision, recall, fscore = evaluate(identified_matches)

print("Precision:", precision)
print("Recall:", recall)
print("Fscore:", fscore)
   

Precision: 0.9056603773584906
Recall: 0.9056603773584906
Fscore: 0.9056603773584906


## Submission

Complete the code in A2-2.ipynb, and submit it to the CourSys activity Assignment 2.